ipatests/test_integration/test_hbac_functional.py

```diff
@@ -0,0 +1,3612 @@
+                         defaults to target host)
+        Returns:
+            True if authentication succeeds, False otherwise
+        """
+        if source_host is None:
+            source_host = host
+
+        result = source_host.run_command([
+            'sshpass', '-p', password, 'ssh',
+            '-o', 'StrictHostKeyChecking=no',
+            '-o', 'PasswordAuthentication=yes',
+            '-l', user, host.hostname, 'echo login successful'
+        ], raiseonerr=False)
+
+        if result.returncode == 0 and 'login successful' in result.stdout_text:
+            return True
+        return False
+
+    def ssh_auth_failure(self, user, password, host, source_host=None):
+        """
+        Test SSH authentication failure
+        Args:
+            user: Username to authenticate
+            password: Password for authentication
+            host: Host to connect to (target host)
+            source_host: Host to run command from (optional,
+                         defaults to target host)
+        Returns:
+            True if authentication fails as expected, False otherwise
+        """
+        if source_host is None:
+            source_host = host
+
+        result = source_host.run_command([
+            'sshpass', '-p', password, 'ssh',
+            '-o', 'StrictHostKeyChecking=no',
+            '-o', 'PasswordAuthentication=yes',
+            '-l', user, host.hostname, 'echo login successful'
+        ], raiseonerr=False)
+
+        if (result.returncode != 0
+                or 'login successful' not in result.stdout_text):
+            return True
+        return False
+
+    def ftp_auth_success(self, user, password, target_host, source_host):
+        """
+        Test FTP authentication success
+        Args:
+            user: Username to authenticate
+            password: Password for authentication
+            target_host: Host to connect to (FTP server)
+            source_host: Host to run command from
+        Returns:
+            True if authentication succeeds, False otherwise
+        """
+        ftp_script = (
+            f'printf "user {user} {password}\nquit\n" | '
+            f'ftp -inv {target_host.hostname}'
+        )
+        result = source_host.run_command(
+            ['bash', '-c', ftp_script],
+            raiseonerr=False
+        )
+
+        if (result.returncode == 0
+                and 'Login successful.' in result.stdout_text):
+            return True
+        return False
+
+    def ftp_auth_failure(self, user, password, target_host, source_host):
+        """
+        Test FTP authentication failure
+        Args:
+            user: Username to authenticate
+            password: Password for authentication
+            target_host: Host to connect to (FTP server)
+            source_host: Host to run command from
+        Returns:
+            True if authentication fails as expected, False otherwise
+        """
+        ftp_script = (
```

Comment:
You can simplify this method with `return not self.ftp_auth_success(user, password, target_host, source_host)`